infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/LogReplicationSinkManager.java

```diff
@@ -430,13 +430,34 @@ private void completeSnapshotApply(LogReplication.LogReplicationEntryMsg entry)
 
         processSnapshotSyncApplied(entry);
 
-        rxState = RxState.LOG_ENTRY_SYNC;
-        logEntrySinkBufferManager = new LogEntrySinkBufferManager(ackCycleTime, ackCycleCnt, bufferSize,
-                logReplicationMetadataManager.getLastProcessedLogEntryBatchTimestamp(), this);
-        logEntryWriter.reset(entry.getMetadata().getSnapshotTimestamp(), entry.getMetadata().getSnapshotTimestamp());
+        // TODO V2: revisit this when increasing the number of threads in logReplicationServer. (fix in PR 3750)
+        // snapshot_Start and completeSnapshotApply is executed by different threads, and they race on updating rxState.
+        // Consider this scenario: Thread1 is working on a snapshot apply with baseSnapshotTimestamp T1 and comes here
```

Comment:
not sure if this scenario is applicable now.. we do not seem to validate anything in processSnapshotStart() based on the Rx State.

Also, there is no synchronized block in processSnapshotStart().  Let me know if I am missing something.

Reply:
Thanks for catching that processSnapshotStart() is not synchronized!
The changes here is for protecting the update of rxState. 
The rxState is validated in receivedValidMessage()...

we do not compare anything while assigning a value to rxState, but since there are 2 threads, we want the some of the comparisons and assignments in this block to use the latest value (cache coherency issue)...which is guaranteed when "synchronized" is used.